models/taskRequests.js

```diff
@@ -0,0 +1,142 @@
+const { TASK_REQUEST_STATUS } = require("../constants/taskRequests");
+const { TASK_STATUS } = require("../constants/tasks");
+const { userState } = require("../constants/userStatus");
+const firestore = require("../utils/firestore");
+const taskRequestsCollection = firestore.collection("taskRequests");
+const tasksModel = require("./tasks");
+const userModel = require("./users");
+const userStatusModel = require("./userStatus");
+
+/**
+ * Fetch all task requests
+ *
+ * @return {Object}
+ */
+const fetchTaskRequests = async () => {
+  try {
+    const taskRequests = await taskRequestsCollection.get();
+
+    const taskRequestsData = [];
+
+    taskRequests.forEach((taskRequest) => taskRequestsData.push({ id: taskRequest.id, ...taskRequest.data() }));
+
+    return taskRequestsData;
+  } catch (err) {
+    logger.error("error fetching tasks", err);
+    throw err;
+  }
+};
+
+/**
+ * Creates a task request
+ *
+ * @param taskId { string }: id of task request
+ * @return {Promise<{taskRequest: Object}>}
+ */
+const createTaskRequest = async (taskId, userId) => {
+  try {
+    const taskRequest = await taskRequestsCollection.doc(taskId).get();
+    const taskRequestData = taskRequest.data();
+
+    const { userExists, user } = await userModel.fetchUser({ userId });
+
+    if (!userExists) {
+      return { userDoesNotExists: true };
+    }
+
+    if (taskRequestData) {
+      const currentRequestors = taskRequestData.requestors;
+      if (currentRequestors.length >= 1) {
+        const requestorExists = !!currentRequestors.find((requestor) => requestor === user.id);
+        if (requestorExists) {
+          return { requestorExists };
+        }
+
+        const updatedRequestors = [...currentRequestors, user.id];
+        await taskRequestsCollection.doc(taskId).update({ requestors: updatedRequestors });
+
+        return {
+          isUpdate: true,
+          requestors: updatedRequestors,
+        };
+      }
+    }
+
+    const { taskData } = await tasksModel.fetchTask(taskId);
+    if (!taskData) {
+      return { taskDoesNotExist: true };
+    }
+
+    const newTaskRequest = {
+      isNoteworthy: taskData.isNoteworthy,
+      priority: taskData.priority,
+      purpose: taskData.purpose || "",
+      requestors: [userId],
+      status: TASK_REQUEST_STATUS.WAITING,
+      title: taskData.title,
+      type: taskData.type,
+    };
+
+    await taskRequestsCollection.doc(taskId).set(newTaskRequest);
```

Comment:
How your creating/getting `taskId`?
Is `taskId` unique enough?

Reply:
For creating a new document preferred method is `add`, not use `set` because of possible conflicts of `id` for a document.

Reply:
The `taskId` is the id of the task for which the user want to create task request for.
In the current setup `taskId` and the id of `taskRequest` document is going to be same, and `set` is only used for cases in which a new `taskRequest` object needs to be created.